### PR TITLE
Stylistic updates to the devdocs for consistency

### PR DESCRIPTION
**Pull request by ararslan**
As an accompaniment to #21463 and its predecessor #21456, this PR seeks to update the stylistic conventions used in the dev docs. In particular this converts code to use `where` and `struct` as applicable, and it applies more consistent Markdown formatting across the board.

CI skipped, remove from commit message when merging.

### CODE REVIEW

File: doc/src/devdocs/stdio.md [hunk 1]
```diff
@@ -2,37 +2,37 @@
 
 ## Libuv wrappers for stdio
 
-`julia.h` defines [libuv](http://docs.libuv.org) wrappers for the `stdio.h` streams:
+julia.h defines [libuv](http://docs.libuv.org) wrappers for the stdio.h streams:
```
Comment (tkelman) [inline doc/src/devdocs/stdio.md]: why remove highlighting here?

File: doc/src/devdocs/stdio.md [hunk 2]
```diff
@@ -2,37 +2,37 @@
...
 ... and corresponding output functions:
 
-```
+```c
 int jl_printf(uv_stream_t *s, const char *format, ...);
 int jl_vprintf(uv_stream_t *s, const char *format, va_list args);
 ```
 
-These `printf` functions are used by `julia/{src,ui}/*.c` wherever stdio is needed to ensure that
+These `printf` functions are used by the .c files in the src/ and ui/ wherever stdio is needed to ensure that
```
Comment (tkelman) [inline doc/src/devdocs/stdio.md]: in the ... folders, and should probably keep highlighting

File: doc/src/devdocs/stdio.md (hunk 1 above)
Comment (ararslan) [inline doc/src/devdocs/stdio.md]: I tried to get things into a consistent style across the docs, and part of that is to use code formatting for code and removing the code formatting from file names and paths that are not code. It provides a better visual distinction IMO, and that's more common in other parts of the docs.

File: doc/src/devdocs/stdio.md [hunk 3]
```diff
@@ -2,37 +2,37 @@
...
 
-```
+```c
 void jl_safe_printf(const char *str, ...);
 ```
 
 ## Interface between JL_STD* and Julia code
 
-[`Base.STDIN`](@ref), [`Base.STDOUT`](@ref) and [`Base.STDERR`](@ref) are bound to the `JL_STD*`[libuv](http://docs.libuv.org)
+[`Base.STDIN`](@ref), [`Base.STDOUT`](@ref) and [`Base.STDERR`](@ref) are bound to the `JL_STD*` libuv
```
Comment (tkelman) [inline doc/src/devdocs/stdio.md]: don't remove the link here, or the highlighting in the rest of this file

Comment (ararslan) [inline doc/src/devdocs/stdio.md]: libuv is linked earlier in the file, so it seemed redundant here.

File: doc/src/devdocs/stdio.md (hunk 1 above)
Comment (tkelman) [inline doc/src/devdocs/stdio.md]: file names that aren't links look a bit strange in a non monospace font imo

Comment (ararslan) [inline doc/src/devdocs/stdio.md]: I sympathize but I still prefer this style. Consider this line, taken from later in the file:

> There is a hack in ios.h that makes the `ios_t.bm` field line up with the `uv_stream_t.type`

Having `ios.h` formatted as code makes it look like it could be an object, as with the nearby `ios_t.bm`. If you stick to having file names unformatted, it's clearer that it isn't an object.

Comment (tkelman) [inline doc/src/devdocs/stdio.md]: Maybe, but anywhere there are wildcards it comes out very wrong without code highlighting. It is an object of sorts, a file - not a normal english prose word.

**Review by StefanKarpinski — APPROVED** [review]
LGTM

File: doc/src/devdocs/stdio.md (hunk 1 above)
Comment (ararslan) [inline doc/src/devdocs/stdio.md]: > anywhere there are wildcards it comes out very wrong without code highlighting

? Line 84 in this file renders fine

> it is an object of sorts

Yes, but not in the same context as the objects that are being referenced around it

Comment (tkelman) [inline doc/src/devdocs/stdio.md]: No it doesn't, it makes the rest of the file incorrectly italic.

Comment (ararslan) [inline doc/src/devdocs/stdio.md]: screen shot 2017-04-20 at 7 44 36 pm

Comment (tkelman) [inline doc/src/devdocs/stdio.md]: The plain text highlighting gets screwed up by it. Isn't going to look right to me showing file names as if they're prose, but if that's the more consistent direction to take things in, we should make that decision and write it down in the documentation standards since it's going to get inconsistent very quickly if people don't consciously think about this.

Comment (ararslan) [inline doc/src/devdocs/stdio.md]: > The plain text highlighting gets screwed up by it.

What plain text highlighting do you mean?

> it's going to get inconsistent very quickly

It already is, which is why I made these changes. Making concrete decisions about formatting documentation and documenting them sounds like a very good idea to me.

Comment (tkelman) [inline doc/src/devdocs/stdio.md]: > What plain text highlighting do you mean?

https://github.com/JuliaLang/julia/pull/21464/files#diff-6b348b7d7ac9e34f6812f75263ac4e15 or https://github.com/JuliaLang/julia/blame/1c7a5c3df1f2ec06ad60ac445b4ae93b3e10aefc/doc/src/devdocs/stdio.md#L84

Comment (ararslan) [inline doc/src/devdocs/stdio.md]: Since it only affects the GitHub web UI and not how the final page renders, that doesn't seem like an issue to me...

File: doc/src/devdocs/init.md [hunk 4]
```diff
@@ -49,183 +49,183 @@ jl_int32_type = jl_new_bitstype(jl_symbol("Int32"),
...
+to execute [boot.jl](https://github.com/JuliaLang/julia/blob/master/base/boot.jl). <!-- TODO – drill
+down into eval? -->
 
-[jl_get_builtin_hooks()](https://github.com/JuliaLang/julia/blob/master/src/init.c) initialises
-global C pointers to Julia globals defined in `boot.jl`.
+[`jl_get_builtin_hooks()`](https://github.com/JuliaLang/julia/blob/master/src/init.c) initializes
+global C pointers to Julia globals defined in boot.jl.
 
-[jl_init_box_caches()](https://github.com/JuliaLang/julia/blob/master/src/alloc.c) pre-allocates
+[`jl_init_box_caches()`](https://github.com/JuliaLang/julia/blob/master/src/alloc.c) pre-allocates
```
Comment (tkelman) [inline doc/src/devdocs/init.md]: this link is dead, now in datatype.c

File: doc/src/devdocs/eval.md [hunk 5]
```diff
@@ -4,9 +4,9 @@ One of the hardest parts about learning how the Julia Language runs code is lear
 the pieces work together to execute a block of code.
 
 Each chunk of code typically makes a trip through many esoteric acronyms such as (in no particular
```
Comment (nalimilan) [inline doc/src/devdocs/eval.md]: BTW, not all of these are acronyms.

Comment (ararslan) [inline doc/src/devdocs/eval.md]: Yeah, I noticed that when I was going through here. Perhaps I should change it to:
> Each chunk of code typically makes a trip through many steps with potentially unfamiliar names, such as (in no particular order) ...

Calling C++ "esoteric" is a little odd IMO. 😛

File: doc/src/devdocs/debuggingtips.md [hunk 6]
```diff
@@ -234,9 +234,9 @@ process)
...
 ## Mozilla's Record and Replay Framework (rr)
 
-Julia now works out of the box with [rr,](http://rr-project.org/) the lightweight recording and
+Julia now works out of the box with [rr](http://rr-project.org/), the lightweight recording and
 deterministic debugging framework from Mozilla. This allows you to replay the trace of an execution
 deterministically.  The replayed execution's address spaces, register contents, syscall data etc
 are exactly the same in every run.
 
-A recent version of `rr` (3.1.0 or higher) is required.
+A recent version of rr (3.1.0 or higher) is required.
```
Comment (tkelman) [inline doc/src/devdocs/debuggingtips.md]: `rr` isn't a word, I think we should keep it highlighted

File: doc/src/devdocs/stdio.md (hunk 1 above)
Comment (tkelman) [inline doc/src/devdocs/stdio.md]: > removing the code formatting from file names and paths that are not code. It provides a better visual distinction IMO, and that's more common in other parts of the docs

I'm still not seeing this. Grepping for `\.jl` or `\.c` under docs, far more of the hits that aren't part of a link are code highlighted than not.

File: doc/src/devdocs/debuggingtips.md (hunk 6 above)
Comment (ararslan) [inline doc/src/devdocs/debuggingtips.md]: It's a name, just like how Julia, R, etc. aren't highlighted

Comment (tkelman) [inline doc/src/devdocs/debuggingtips.md]: `julia` when referring to the command-line executable usually is

Comment (ararslan) [inline doc/src/devdocs/debuggingtips.md]: Right, but it seems to me that here it's referring to the name. Just as we would usually say "a recent version of Julia (0.5 or higher) is required," rather than "version of `julia` (0.5 or higher)."

File: doc/src/devdocs/stdio.md (hunk 1 above)
Comment (ararslan) [inline doc/src/devdocs/stdio.md]: Okay, well either way it's inconsistent. I'm inclined to just merge this then make everything consistent in another PR.

Comment (tkelman) [inline doc/src/devdocs/stdio.md]: This is making it less consistent, diverging in this file from what is more commonly used in the rest of the docs.

File: doc/src/devdocs/debuggingtips.md (hunk 6 above)
Comment (tkelman) [inline doc/src/devdocs/debuggingtips.md]: proper nouns referring to the name of something would usually be capitalized, which makes the distinction clear

File: doc/src/devdocs/stdio.md (hunk 1 above)
Comment (ararslan) [inline doc/src/devdocs/stdio.md]: Paths and file names are now highlighted.

File: doc/src/devdocs/init.md [hunk 7]
```diff
@@ -49,183 +49,183 @@ jl_int32_type = jl_new_bitstype(jl_symbol("Int32"),
...
 
-`_julia_init()` then returns [back to main() in julia/ui/repl.c](https://github.com/JuliaLang/julia/blob/master/ui/repl.c)
-and main() calls `true_main(argc, (char**)argv)`.
+`_julia_init()` then returns [back to `main()` in julia/ui/repl.c](https://github.com/JuliaLang/julia/blob/master/ui/repl.c)
+and `main()` calls `true_main(argc, (char**)argv)`.
 
 !!! sidebar "sysimg"
     If there is a sysimg file, it contains a pre-cooked image of the `Core` and `Main` modules (and
-    whatever else is created by `boot.jl`). See [Building the Julia system image](@ref).
+    whatever else is created by boot.jl). See [Building the Julia system image](@ref).
```
Comment (tkelman) [inline doc/src/devdocs/init.md]: put this back

File: doc/src/devdocs/stdio.md [hunk 8]
```diff
@@ -60,43 +60,43 @@ stream.jl: function write(s::IO, p::Ptr, nb::Integer)
...
+any particular piece of code is reachable before initialization is complete.
 
 ## Legacy ios.c library
 
-The `julia/src/support/ios.c` library is inherited from [femtolisp](https://github.com/JeffBezanson/femtolisp).
+The `src/support/ios.c` library is inherited from [femtolisp](https://github.com/JeffBezanson/femtolisp).
 It provides cross-platform buffered file IO and in-memory temporary buffers.
 
-`ios.c` is still used by:
+ios.c is still used by:
```
Comment (tkelman) [inline doc/src/devdocs/stdio.md]: put back
